src/smtsolvers/GhostSMTSolver.cpp

```diff
@@ -0,0 +1,250 @@
+//
+// Created by prova on 14.11.19.
+//
+
+#include "GhostSMTSolver.h"
+#include <utility>
+
+bool GhostSMTSolver::isGhost(Lit l)
+{
+    if (!theory_handler.isTheoryTerm(var(l))) return false;
+    vec<CRef> &appearances = thLitToClauses[toInt(l)];
+    int i;
+    for (i = 0; i < appearances.size(); i++) {
+        if (!satisfied(ca[appearances[i]])) {
+            std::swap(appearances[0], appearances[i]);
+            break;
+        }
+    }
+    return i == appearances.size();
+}
+
+void GhostSMTSolver::newDecisionLevel()
+{
+    SimpSMTSolver::newDecisionLevel();
+    ghostTrailLim.push(ghostTrail.size());
+}
+
+void GhostSMTSolver::cancelUntil(int level) {
+    int prev_dl = decisionLevel();
+    SimpSMTSolver::cancelUntil(level);
+    if (prev_dl > level) {
+        for (int c = ghostTrail.size() - 1; c >= ghostTrailLim[level]; c--) {
+            insertVarOrder(ghostTrail[c]);
+        }
+        ghostTrail.shrink(ghostTrail.size() - ghostTrailLim[level]);
+        ghostTrailLim.shrink(ghostTrailLim.size() - level);
+    }
+    assert(ghostTrailLim.size() == trail_lim.size());
+}
+
+void
+GhostSMTSolver::attachClause(CRef in_clause)
+{
+    SimpSMTSolver::attachClause(in_clause);
+
+    if (in_clause == CRef_Undef)
+        return;
+
+    const Clause& c = ca[in_clause];
+    if (c.learnt())
+        return;
+
+    for (int i = 0; i < c.size(); i++) {
+        Lit l = c[i];
+        if (theory_handler.isTheoryTerm(var(l))) {
+            int idx = toInt(l);
+            assert(idx < thLitToClauses.size());
+            thLitToClauses[idx].push(in_clause);
+        }
+    }
+}
+
+void
+GhostSMTSolver::detachClause(CRef cr, bool strict)
+{
+    SimpSMTSolver::detachClause(cr, strict);
+    const Clause& c = ca[cr];
+    if (c.learnt())
+        return;
+
+    assert(false);
+    for (int i = 0; i < c.size(); i++) {
```

Comment:
Why do we need code after assert(false)?
If it is not an error than there should not be assert(false), if it is error than probably something is wrong with the solver and we should just abort.

Reply:
True, I removed the code, thanks!